### PR TITLE
:sparkles: client: Generate apply configurations

### PR DESCRIPTION
**Pull request by astefanutti**
## Summary

This PR enables the generation of apply configurations for kcp APIs, so they can be used for server-side apply requests.

Requires https://github.com/kubernetes/kubernetes/pull/114920.

### CODE REVIEW

File: hack/update-codegen-clients.sh [hunk 1]
```diff
@@ -27,7 +27,42 @@ BOILERPLATE_HEADER="$( pwd )/hack/boilerplate/boilerplate.go.txt"
...
+  --input-dirs github.com/kcp-dev/kcp/pkg/apis/topology/v1alpha1 \
+  --input-dirs github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1 \
+  --input-dirs github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1 \
+  --input-dirs k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/version,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1 \
+  --output-package github.com/kcp-dev/kcp/pkg/client/applyconfiguration \
+  --go-header-file ./hack/../hack/boilerplate/boilerplate.generatego.txt \
+  --output-base "${SCRIPT_ROOT}" \
+  --trim-path-prefix github.com/kcp-dev/kcp
+
+"$GOPATH"/bin/client-gen \
```
Comment (stevekuznetsov) [inline hack/update-codegen-clients.sh:48]: What does this call to `client-gen` do?

**Review by stevekuznetsov — COMMENTED** [review]
Looks awesome!

File: hack/update-codegen-clients.sh (hunk 1 above)
Comment (astefanutti) [reply hack/update-codegen-clients.sh:48]: The upstream `generate-groups.sh` script does not support passing the `apply-configuration-package` option to `client-gen`, so this call does what `generate-groups.sh "client"` would do, but with the extra `apply-configuration-package` passed to `client-gen`. Upstream `code-generator` scripts completely lack support for `applyconfiguration-gen`, so I went for the path of least resistance, and have `code-generator` improved asynchronously.

File: hack/update-codegen-clients.sh [hunk 2]
```diff
@@ -36,7 +71,7 @@ bash "${CODEGEN_PKG}"/generate-groups.sh "deepcopy,client" \
 
 pushd ./pkg/apis
 ${CODE_GENERATOR} \
-  "client:outputPackagePath=github.com/kcp-dev/kcp/pkg/client,apiPackagePath=github.com/kcp-dev/kcp/pkg/apis,singleClusterClientPackagePath=github.com/kcp-dev/kcp/pkg/client/clientset/versioned,headerFile=${BOILERPLATE_HEADER}" \
+  "client:outputPackagePath=github.com/kcp-dev/kcp/pkg/client,apiPackagePath=github.com/kcp-dev/kcp/pkg/apis,singleClusterClientPackagePath=github.com/kcp-dev/kcp/pkg/client/clientset/versioned,singleClusterApplyConfigurationsPackagePath=github.com/kcp-dev/kcp/pkg/client/applyconfiguration,headerFile=${BOILERPLATE_HEADER}" \
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply hack/update-codegen-clients.sh:74]: Right. Let me do it 👍🏼.

File: hack/update-codegen-clients.sh (hunk 1 above)
Comment (astefanutti) [reply hack/update-codegen-clients.sh:48]: Upstream PR: kubernetes/kubernetes#114987

File: hack/update-codegen-clients.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (astefanutti) [reply hack/update-codegen-clients.sh:74]: Updated. PTAL.

**Pull request merged**